File: cs/XDust/XDust/Context.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace XDust
{
    public class Context : IEnumerable<String>, IScriptable
    {
        public Context(Object head, Object tail, Dictionary<String, Object> parameters)
        {
            this.Parameters = new Dictionary<String, Object>();
            if (null != parameters)
            {
                foreach (var name in parameters.Keys)
                {
                    this.Parameters[name] = parameters[name];
                }
            }
            this.Tail = Scriptable.From(tail);
        }

        public Context Head { get; set; }
        public IScriptable Tail { get; set; }
        public Dictionary<String, Object> Parameters { get; private set; }

        public Object this[String name]
        {
            get
            {
                return this.Get(name);
            }
        }

        public Object Get(String name)
        {
            if (name == ".")
            {
                return this.Value;
            }
            else if (this.Tail.HasKey(name))
            {
                return this.Tail[name];
            }
            else if (this.Parameters.ContainsKey(name))
            {
                return this.Parameters[name];
            }
            else if (null != this.Head)
            {
                return this.Head.Get(name);
            }
            else
            {
                return null;
            }
        }

        public void Update(Object other)
        {
            if (null != other)
            {
                this.Tail = Scriptable.From(other);
                if (other is Context)
                {
                    foreach (var param in ((Context)other).Parameters)
                    {
                        this.Parameters[param.Key] = param.Value;
                    }
                }
            }
        }

        #region IEnumerable<string> Members

        public IEnumerator<String> GetEnumerator()
        {
            if (this.Tail.IsValue)
            {
                throw new InvalidOperationException();
            }
            else
            {
                return this.Tail.GetEnumerator();
            }
        }

        #endregion

        #region IEnumerable Members

        System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator()
        {
            return this.GetEnumerator();
        }

        #endregion

        #region IEnumerable<string> Members

        IEnumerator<string> IEnumerable<string>.GetEnumerator()
        {
            throw new NotImplementedException();
        }

        #endregion

        #region IScriptable Members

        public object Value
        {
            get
            {
                if (null != this.Tail)
                {
                    return this.Tail.Value;
                }
                else if (null != this.Head)
                {
                    return this.Head.Value;
                }
                else
                {
                    return null;
                }
            }
        }

        public bool IsValue
        {
            get
            {
                if (null != this.Tail)
                {
                    return this.Tail.IsValue;
                }
                else if (null != this.Head)
                {
                    return this.Head.IsValue;
                }
                else
                {
                    return false;
                }
            }
        }

        public bool HasKey(string name)
        {
            if (null != this.Parameters && this.Parameters.ContainsKey(name))
            {
                return true;
            }
            else if (null != this.Tail && this.Tail.HasKey(name))
            {
                return true;
            }
            else if (null != this.Head && this.Head.HasKey(name))
            {
                return true;
            }
            else
            {
                return false;
            }
        }

        #endregion
    }
}

File: java/src/xdust-mvn/src/main/java/com/heydanno/xdust/Context.java
package com.heydanno.xdust;

import java.io.Serializable;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

/**
 * Defines a data context used by a template
 */
public class Context implements IScriptable, Serializable {

	private static final long serialVersionUID = -5834248967445485253L;

	/**
	 * Constructor
	 * 
	 * @param head
	 *            The parent context (if any)
	 * @param tail
	 *            The current data object (if any)
	 * @param parameters
	 *            Variables to set for the current context
	 */
	public Context(Object head, Object tail, Map<String, ?> parameters) {
		if (null != parameters) {
			for (String key : parameters.keySet()) {
				this.getParameters().put(key, parameters.get(key));
			}
		}
		this.tail = Scriptable.from(tail);
	}

	private Context head;
	private IScriptable tail;
	private Map<String, Object> parameters;

	/**
	 * Gets the parent context
	 * 
	 * @return The parent context
	 */
	public Context getHead() {
		return this.head;
	}

	/**
	 * Sets the parent context
	 * 
	 * @param value
	 *            The parent context
	 */
	public void setHead(Context value) {
		this.head = value;
	}

	/**
	 * Gets the current context's model
	 * 
	 * @return The current context's model
	 */
	public IScriptable getTail() {
		return this.tail;
	}

	/**
	 * Sets the current context's model
	 * 
	 * @param value
	 *            The current context's model
	 */
	public void setTail(IScriptable value) {
		this.tail = value;
	}

	/**
	 * Gets the variables associated with the current context
	 * 
	 * @return The variables associated with the current context
	 */
	public Map<String, Object> getParameters() {
		if (null == this.parameters) {
			this.parameters = new HashMap<String, Object>();
		}
		return this.parameters;
	}

	/**
	 * Gets a raw value represented by the current context (if any)
	 * 
	 * @return The value
	 */
	public Object getValue() {
		if (null != this.tail) {
			return this.tail.getValue();
		} else if (null != this.head) {
			return this.head.getValue();
		} else {
			return null;
		}
	}

	/**
	 * Determines whether or not the current context represents a raw value
	 * 
	 * @return True if the current context is a raw value
	 */
	public boolean isValue() {
		if (null != this.tail) {
			return this.tail.isValue();
		} else if (null != this.head) {
			return this.head.isValue();
		} else {
			return false;
		}
	}

	/**
	 * Gets a value stored under a key from the current context, coming from
	 * either its parameters, its model, or its parent
	 * 
	 * @param name
	 *            The key under which the value is stored
	 * @return The value of that key
	 */
	public Object get(String name) {
		if (".".equals(name)) {
			return this.getValue();
		} else if (this.tail.hasKey(name)) {
			return this.tail.get(name);
		} else if (this.getParameters().containsKey(name)) {
			return this.getParameters().get(name);
		} else if (null != this.head) {
			return this.head.get(name);
		} else {
			return null;
		}
	}

	/**
	 * Determines whether or not there's a value associated with a key in this
	 * context chain
	 * 
	 * @param name
	 *            The key
	 * @return True if there's a value for that key in this context chain
	 */
	public boolean hasKey(String name) {
		if (null != this.parameters && this.parameters.containsKey(name)) {
			return true;
		} else if (null != this.tail && this.tail.hasKey(name)) {
			return true;
		} else if (null != this.head && this.head.hasKey(name)) {
			return true;
		} else {
			return false;
		}
	}

	/**
	 * Allows looping through this context's model
	 * 
	 * @return An iterator
	 */
	public Iterator<String> iterator() {
		if (this.tail.isValue()) {
			return null;
		} else {
			return this.tail.iterator();
		}
	}

	/**
	 * Indicates whether or not this context's model is a list
	 * 
	 * @return True if the context's model is a list
	 */
	public boolean isList() {
		return null != this.tail && this.tail.isList();
	}

	/**
	 * Indicates whether or not this context's model follows the JavaScript
	 * rules for truthy/falsey values
	 * 
	 * @return True if the context's model meets the JavaScript criteria for
	 *         true
	 */
	public boolean isTruthy() {
		return null != this.tail && this.tail.isTruthy();
	}

	/**
	 * Gets an iterable representation of this context's model
	 * 
	 * @return This context's model as a List
	 */
	public List<Object> asList() {
		return this.isList() ? this.tail.asList() : null;
	}

}
